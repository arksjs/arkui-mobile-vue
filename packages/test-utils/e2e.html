<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
