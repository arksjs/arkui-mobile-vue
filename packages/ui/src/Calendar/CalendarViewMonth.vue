<template>
  <div class="fx-calendar-view_month-caption">
    {{ month.caption }}
  </div>
  <div
    class="fx-calendar-view_days"
    :data-index="monthIndex"
    @click="onDaysClick"
  >
    <div
      class="fx-calendar-view_day"
      :class="{
        disabled: day.state === 'disabled',
        selected:
          day.state === 'selected' ||
          day.state === 'startSelected' ||
          day.state === 'endSelected',
        'in-range': mode === 'range' && day.state === 'selected'
      }"
      v-for="(day, dayIndex) in month.days"
      :key="dayIndex"
      :data-index="dayIndex"
    >
      <span
        v-if="day.topText"
        class="fx-calendar-view_day-top-text"
        :class="{ highlight: day.topHighlight }"
        >{{ day.topText }}</span
      >
      {{ day.text }}
      <span
        v-if="day.bottomText"
        class="fx-calendar-view_day-bottom-text"
        :class="{ highlight: day.bottomHighlight }"
        >{{ day.bottomText }}</span
      >
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, PropType } from 'vue'
import type { OnClick } from '../helpers/types'
import type { Mode, Month } from './types'

export default defineComponent({
  name: 'fx-calendar-view-month',
  props: {
    month: {
      type: Object as PropType<Month>,
      required: true
    },
    monthIndex: {
      type: Number,
      required: true
    },
    onDaysClick: {
      type: Function as PropType<OnClick>,
      required: true
    },
    mode: {
      type: String as PropType<Mode>,
      required: true
    }
  }
})
</script>
