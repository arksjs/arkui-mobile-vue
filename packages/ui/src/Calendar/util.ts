import type { CalendarDetail, CalendarSelectorDetail, Mode } from './types'
import { getNumber, isInNumberRange } from '../helpers/util'
import dayjs from '../helpers/day'

const DEFAULT_MONTH_RANGE = 6
export const MODE_NAMES: Mode[] = ['single', 'range']

export function getDefaultSourceDetail(): CalendarDetail {
  return {
    value: [],
    label: '',
    valueArray: [],
    rangeCount: 0
  }
}

export function getSourceDetail(
  newDetail: CalendarSelectorDetail
): CalendarDetail {
  return Object.assign(newDetail.source, {
    valueArray: newDetail.valueArray,
    rangeCount: newDetail.rangeCount
  })
}

export function getFirstDayOfWeek(firstDayOfWeek?: number | string) {
  const num = getNumber(firstDayOfWeek, 0)

  return isInNumberRange(num, 0, 6) ? Math.round(num) : 0
}

export const getTimeByDate = (date: Date) =>
  dayjs(date).startOf('day').valueOf()

export const getMinTime = () => dayjs().startOf('day').valueOf()

export const getMaxTime = (minTime: number) =>
  dayjs(minTime).startOf('day').add(DEFAULT_MONTH_RANGE, 'month').valueOf()

export const getViewBodyTitleStyles = (titleY: number | null) => ({
  transform: `translate3d(0px, ${
    titleY == null ? '-100%' : titleY + 'px'
  }, 0px)`
})
