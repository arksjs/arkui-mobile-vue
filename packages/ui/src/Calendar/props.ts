import type { PropType } from 'vue'
import { isInNumberRange, isInteger } from '../helpers/util'
import type {
  DayHandler,
  Mode,
  ValueFormatter,
  ValueParser,
  CalendarDetail
} from './types'
import { createEnumsValidator } from '../helpers/validator'
import { isPickerDetail } from '../Picker/props'
import type { SelectorModelValue } from '../SelectorField/types'
import { MODE_NAMES } from './util'

export const commonProps = {
  modelValue: {
    type: [Date, String, Number, Array] as PropType<SelectorModelValue>
  },
  minDate: {
    type: Date
  },
  maxDate: {
    type: Date
  },
  initialMode: {
    type: String as PropType<Mode>,
    validator: createEnumsValidator(MODE_NAMES),
    default: MODE_NAMES[0]
  },
  allowSameDay: {
    type: Boolean,
    default: false
  },
  maxRange: {
    type: Number,
    default: Infinity
  },
  dayHandler: {
    type: Function as PropType<DayHandler>
  },
  firstDayOfWeek: {
    validator: (val: number) => {
      return isInteger(val) && isInNumberRange(val, 0, 6)
    },
    default: 0
  },
  formatter: {
    type: Function as PropType<ValueFormatter>
  },
  parser: {
    type: Function as PropType<ValueParser>
  }
}

export const calendarDetailValidator = (payload: CalendarDetail) =>
  isPickerDetail(payload) &&
  typeof payload.rangeCount === 'number' &&
  Array.isArray(payload.valueArray)
