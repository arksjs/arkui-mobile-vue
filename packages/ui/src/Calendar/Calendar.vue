<template>
  <div class="ta-calendar" :class="{ disabled }" ref="root">
    <SelectorField
      :label="fieldLabel"
      :value="fieldValue"
      :disabled="disabled"
      :name="name"
      :placeholder="placeholder"
      @fieldClick="onFieldClick"
    />
    <CalendarPopup
      :modelValue="modelValue"
      :minDate="minDate"
      :maxDate="maxDate"
      :initialMode="initialMode"
      :allowSameDay="allowSameDay"
      :maxRange="maxRange"
      :dayHandler="dayHandler"
      :firstDayOfWeek="firstDayOfWeek"
      :title="placeholder"
      :formatter="formatter"
      :parser="parser"
      :showConfirm="showConfirm"
      :showClose="showClose"
      v-model:visible="popupVisible"
      v-if="isInitPopup"
      @confirm="onConfirm"
      ref="popupRef"
    />
  </div>
</template>

<script lang="ts">
import { defineComponent, ref, shallowRef, watch } from 'vue'
import { SelectorField } from '../SelectorField'
import CalendarPopup from './CalendarPopup.vue'
import { commonProps } from './props'
import type {
  CalendarSelectorDetail,
  CalendarPopupRef,
  CalendarEmits
} from './types'
import { useHandlers } from '../Calendar/use-calendar'
import { cloneDetail, isSameValue } from '../Picker/util'
import { pickerEmits, pickerProps } from '../Picker/props'
import type { PropsToEmits } from '../helpers/types'

export default defineComponent({
  name: 'ta-calendar',
  components: { SelectorField, CalendarPopup },
  props: {
    ...commonProps,
    ...pickerProps,
    showConfirm: {
      type: Boolean,
      default: false
    },
    showClose: {
      type: Boolean,
      default: false
    }
  },
  emits: { ...pickerEmits } as PropsToEmits<CalendarEmits>,
  setup(props, ctx) {
    const { emit } = ctx
    const isInitPopup = ref(false)
    const popupVisible = ref(true)
    const fieldLabel = ref('')
    const fieldValue = ref('')
    const popupRef = shallowRef<CalendarPopupRef | null>(null)
    const root = shallowRef<HTMLElement | null>(null)

    const { formatter, parser, getDefaultDetail } = useHandlers(props)

    let detail: CalendarSelectorDetail = getDefaultDetail()

    function getPopupDetail() {
      return popupRef.value?.getDetail() || getDefaultDetail()
    }

    function updateValue(val: unknown) {
      if (val == null) {
        // 解决 formily 强制null的问题
        return
      }

      updateDetail(formatter(parser(val)))
    }

    function updateDetail(newDetail: CalendarSelectorDetail) {
      detail = newDetail

      fieldLabel.value = newDetail.label
      fieldValue.value =
        detail.value != null
          ? detail.valueArray.map(v => v.join('-')).join(',')
          : ''
    }

    function onFieldClick() {
      if (!props.disabled) {
        if (!isInitPopup.value) {
          isInitPopup.value = true
        } else {
          popupVisible.value = true
        }
      }
    }

    function getDetail() {
      return cloneDetail(detail)
    }

    function onConfirm() {
      const newDetail = getPopupDetail()

      if (isSameValue(detail.value, newDetail.value)) {
        return
      }

      updateDetail(newDetail)

      emit('update:modelValue', getDetail().value)
      emit('change', getDetail().value)
    }

    watch(
      () => props.modelValue,
      val => updateValue(val),
      {
        deep: true,
        immediate: true
      }
    )

    watch([isInitPopup, popupVisible], ([isInit, visible]) => {
      if (isInit && visible) {
        emit('focus')
      } else if (!visible) {
        emit('blur')
      }
    })

    return {
      root,
      isInitPopup,
      popupVisible,
      fieldLabel,
      fieldValue,
      popupRef,
      onFieldClick,
      onConfirm
    }
  }
})
</script>
