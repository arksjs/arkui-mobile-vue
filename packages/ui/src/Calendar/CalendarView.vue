<template>
  <div class="fx-calendar-view">
    <div class="fx-calendar-view_header">
      <div class="fx-calendar-view_weekdays">
        <span
          class="fx-calendar-view_weekday"
          :class="{ highlight: weekDay === '0' || weekDay === '6' }"
          v-for="weekDay in weekDays"
          :key="weekDay"
          >{{ locale.calendarWeekDayTexts[weekDay] }}
        </span>
      </div>
    </div>
    <div class="fx-calendar-view_body" ref="bodyEl">
      <VirtualList
        :ids="months.map(v => v.caption)"
        :itemSize="getItemSize"
        @scroll="onScroll"
      >
        <template #default="{ index }">
          <ViewMonth
            :month="months[index]"
            :monthIndex="index"
            :mode="mode"
            :onDaysClick="onDaysClick"
          />
        </template>
      </VirtualList>
      <div class="fx-calendar-view_month-caption fixed" ref="bodyTitleEl"></div>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, onMounted, reactive, ref, watch } from 'vue'
import dayjs from '../helpers/day'
import type { Dayjs, PropsToEmits } from '../helpers/types'
import { isSameArray } from '../helpers/util'
import { showToast } from '../Toast'
import { commonProps, calendarDetailValidator } from './props'
import {
  printError,
  getMinTime,
  getMaxTime,
  getTimeByDate,
  getFirstDayOfWeek,
  getViewBodyTitleStyles
} from './util'
import type { CalendarViewEmits, DayInfo, Month } from './types'
import { useHandlers } from '../Calendar/use-calendar'
import { pickerViewEmits } from '../Picker/props'
import { useLocale } from '../ConfigProvider/context'
import ViewMonth from './CalendarViewMonth.vue'
import { VirtualList } from '../VirtualList'
import { CSSProperties2CssText, getScrollTop } from '../helpers/dom'

type WeekDay = '0' | '1' | '2' | '3' | '4' | '5' | '6'

interface SelectDay {
  dateString: string
  timestamp: number
  monthIndex: number
  dayIndex: number
  state?: string
}

const defaultWeekDays: WeekDay[] = ['0', '1', '2', '3', '4', '5', '6']

function getDefaultSelectDay() {
  return {
    dateString: '',
    timestamp: 0,
    monthIndex: 0,
    dayIndex: 0
  }
}

export default defineComponent({
  name: 'fx-calendar-view',
  components: { ViewMonth, VirtualList },
  props: { ...commonProps },
  emits: {
    ...pickerViewEmits,
    select: calendarDetailValidator
  } as PropsToEmits<CalendarViewEmits>,
  setup(props, { emit }) {
    const { locale } = useLocale()
    const { formatter, parser, mode } = useHandlers(props)
    const bodyEl = ref<HTMLDivElement>()
    const bodyTitleEl = ref<HTMLDivElement>()

    const weekDays = ref<WeekDay[]>([])
    const months = reactive<Month[]>([])

    let start: SelectDay = getDefaultSelectDay()
    let end: SelectDay = getDefaultSelectDay()

    function getSelectedInfo(timestamp: number): SelectDay | null {
      for (let i = 0; i < months.length; i++) {
        for (let j = 0; j < months[i].days.length; j++) {
          const day = months[i].days[j]

          if (day.state !== 'disabled') {
            if (timestamp === day.timestamp) {
              return {
                dateString: day.dateString,
                timestamp,
                monthIndex: i,
                dayIndex: j
              }
            }
          }
        }
      }

      return null
    }

    function updateValue(val: unknown) {
      return updateOriginalValue(parser(val))
    }

    function updateOriginalValue(timeArr: number[]) {
      if (timeArr.length === 1) {
        timeArr.push(0)
      }

      if (!isSameArray(timeArr, [start.timestamp, end.timestamp])) {
        if (timeArr.length === 0 || timeArr[0] === 0) {
          setSelected('start', null)
          setSelected('end', null)
          updateStates()
        } else if (mode === 'range') {
          const _start = getSelectedInfo(timeArr[0])
          const _end = getSelectedInfo(timeArr[1])

          if (_start && _end) {
            const { rangeCount, hasDisabled } = getRangeInfo(_start, _end)

            if (hasDisabled) {
              printError('The range of "modelValue" contains disabled days.')
            } else if (rangeCount > props.maxRange) {
              printError(
                `The range of "modelValue" contains ${rangeCount} days, no more than ${props.maxRange} days.`
              )
            } else {
              setSelected('start', _start)
              setSelected('end', _end)
              updateStates()
            }
          } else {
            printError(
              'The range of "modelValue" is not in the optional range.'
            )
          }
        } else {
          const select = getSelectedInfo(timeArr[0])

          if (select) {
            start = select
            setSelected('end', null)
            updateStates()
          } else {
            printError(
              'The range of "modelValue" is not in the optional range.'
            )
          }
        }
      }

      return getDetail()
    }

    function setSelected(name: 'start' | 'end', day: SelectDay | null) {
      if (day) {
        name === 'start' ? (start = day) : (end = day)
      } else {
        name === 'start'
          ? (start = getDefaultSelectDay())
          : (end = getDefaultSelectDay())
      }
    }

    function getState(timestamp: number) {
      let state = ''

      if (
        (mode === 'range' &&
          timestamp >= start.timestamp &&
          timestamp <= end.timestamp) ||
        timestamp === start.timestamp
      ) {
        state = 'selected'
      }
      if (mode === 'range' && state == 'selected') {
        if (timestamp === end.timestamp) {
          state = 'endSelected'
        } else if (timestamp === start.timestamp) {
          state = 'startSelected'
        }
      }

      return state
    }

    function getDayInfo(day: Dayjs, extend: { state: string }): DayInfo {
      const dateString = day.format('YYYY-MM-DD')
      const state = extend.state

      let dayInfo: DayInfo = {
        topHighlight: false,
        topText:
          state === 'startSelected'
            ? locale.value.calendarSelectedStartText
            : state === 'endSelected'
            ? locale.value.calendarSelectedEndText
            : '',
        state,
        bottomHighlight: false,
        bottomText: '',
        text: day.date().toString(),
        dateString,
        timestamp: day.valueOf()
      }

      if (props.dayHandler) {
        dayInfo.date = day.toDate()
        dayInfo = props.dayHandler(Object.assign(dayInfo, extend))
        delete dayInfo.date
      }

      if (dayInfo.state === 'disabled' && !extend.state) {
        extend.state = 'disabled'
      }

      return Object.assign(dayInfo, extend, {
        dateString,
        timestamp: day.valueOf()
      })
    }

    function getStartMonth(day: Dayjs) {
      const month: Month = {
        caption: day.format(locale.value.calendarMonthCaption),
        monthString: day.format('YYYY-MM'),
        days: [],
        startDay: day.date()
      }

      let day2 = day.startOf('month')

      // 头部周偏移占位
      for (
        let i = 0, len = day2.day() - getFirstDayOfWeek(props.firstDayOfWeek);
        i < len;
        i++
      ) {
        month.days.push({
          cover: true,
          text: '',
          state: 'disabled',
          dateString: '',
          timestamp: 0
        })
      }

      while (day2.date() < month.startDay) {
        month.days.push(getDayInfo(day2, { state: 'disabled' }))
        day2 = day2.add(1, 'day')
      }

      return month
    }

    function updateWeekDays() {
      let i = getFirstDayOfWeek()
      const newWeekDays: WeekDay[] = []

      let weekDay: WeekDay
      while (newWeekDays.length < 7) {
        weekDay = defaultWeekDays[i]
        newWeekDays.push(weekDay)
        i = (i + 1) % 7
      }

      weekDays.value = newWeekDays
    }

    let minTimestamp = 0
    let maxTimestamp = 0

    function updateOptions() {
      if (props.minDate instanceof Date) {
        minTimestamp = getTimeByDate(props.minDate)
      } else {
        minTimestamp = getMinTime()
      }

      if (props.maxDate instanceof Date) {
        if (props.maxDate.getTime() < minTimestamp) {
          printError(
            'The value of "maxDate" cannot be less than the value of "minDate".'
          )
          maxTimestamp = getMaxTime(minTimestamp)
        } else {
          maxTimestamp = getTimeByDate(props.maxDate)
        }
      } else {
        maxTimestamp = getMaxTime(minTimestamp)
      }

      updateWeekDays()

      const maxDay = dayjs(maxTimestamp)
      const _months = []

      let day = dayjs(minTimestamp)
      let monthKey = day.month()
      let month = getStartMonth(day)

      while (!day.isAfter(maxDay)) {
        if (day.month() !== monthKey) {
          monthKey = day.month()
          _months.push(month)
          month = getStartMonth(day)
        }

        const dayInfo = getDayInfo(day, {
          state: getState(day.valueOf())
        })

        // if (
        //   dayInfo.state === 'startSelected' ||
        //   (mode === 'single' && dayInfo.state === 'selected')
        // ) {
        //   setSelected('start', dayInfo, _months.length, month.days.length)
        // } else if (dayInfo.state === 'endSelected') {
        //   setSelected('end', dayInfo, _months.length, month.days.length)
        // }

        month.days.push(dayInfo)
        day = day.add(1, 'day')
      }

      // 补上最后一个月结尾的天数
      while (day.month() === monthKey) {
        month.days.push(
          getDayInfo(day, {
            state: 'disabled'
          })
        )

        day = day.add(1, 'day')
      }

      _months.push(month)

      months.splice(0, Infinity, ..._months)

      updateBodyFixed(bodyScrollTop)
    }

    function dayInfo2SelectDay(
      day: DayInfo,
      monthIndex: number,
      dayIndex: number
    ): SelectDay {
      return {
        dateString: day.dateString,
        timestamp: day.timestamp,
        state: day.state,
        monthIndex,
        dayIndex
      }
    }

    function onDaysClick(e: Event) {
      const target = e.target as HTMLElement
      let $day: HTMLElement | null = null

      if (target.tagName === 'SPAN') {
        $day = target.parentElement as HTMLElement
      } else if (target !== e.currentTarget) {
        $day = target
      }

      if (!$day) {
        return
      }

      const monthIndex = parseInt(
        (e.currentTarget as HTMLElement).dataset.index as string
      )
      const dayIndex = parseInt($day.dataset.index as string)
      const day = months[monthIndex].days[dayIndex]

      if (day.state === 'disabled') {
        return
      }

      const newDay = dayInfo2SelectDay(day, monthIndex, dayIndex)

      if (mode === 'range') {
        // 范围
        if ((start.dateString && end.dateString) || !start.dateString) {
          setSelected('end', null)
        } else {
          if (
            day.timestamp > start.timestamp ||
            (props.allowSameDay && day.timestamp === start.timestamp)
          ) {
            // 范围
            const { rangeCount, hasDisabled } = getRangeInfo(start, {
              monthIndex,
              dayIndex
            })

            if (!hasDisabled) {
              if (rangeCount > props.maxRange) {
                showToast(
                  locale.value.calendarMaxRangeTips.replace(
                    '{{maxRange}}',
                    props.maxRange.toString()
                  )
                )
              } else {
                setSelected('end', newDay)
                updateStates()
                onSelect()
              }
              return
            }
          }
        }
      } else {
        // 单选
        setSelected('start', newDay)
        updateStates()
        onSelect()
        return
      }

      // range 设置开始时间
      setSelected('start', newDay)
      updateStates()
    }

    function updateStates() {
      for (let i = 0; i < months.length; i++) {
        for (let j = 0; j < months[i].days.length; j++) {
          const day = months[i].days[j]

          if (day.state !== 'disabled') {
            const newState = getState(day.timestamp)

            if (newState !== day.state) {
              months[i].days[j] = getDayInfo(dayjs(day.timestamp), {
                state: newState
              })
            }
          }
        }
      }
    }

    let timeValue = [start.timestamp, end.timestamp]

    function onChange() {
      if (isSameArray([start.timestamp, end.timestamp], timeValue)) {
        return
      }

      timeValue = [start.timestamp, end.timestamp]
      emit('update:modelValue', getDetail().value)
      emit('change', getDetail().value)
    }

    function onSelect() {
      onChange()
      emit('select', getDetail())
    }

    function getDetail() {
      return formatter([start.timestamp, end.timestamp])
    }

    /**
     * 判断所选范围内有没有 disabled
     */
    function getRangeInfo(
      start: { monthIndex: number; dayIndex: number },
      end: { monthIndex: number; dayIndex: number }
    ) {
      let hasDisabled = false
      let rangeCount =
        start.monthIndex === end.monthIndex && start.dayIndex === end.dayIndex
          ? 1
          : 2

      for (let i = start.monthIndex; i <= end.monthIndex; i++) {
        for (
          let j = i === start.monthIndex ? start.dayIndex + 1 : 0,
            len = i === end.monthIndex ? end.dayIndex : months[i].days.length;
          j < len;
          j++
        ) {
          const day = months[i].days[j]

          if (!day.cover) {
            if (day.state === 'disabled') {
              hasDisabled = true
            } else {
              rangeCount++
            }
          }
        }
      }

      return {
        hasDisabled,
        rangeCount
      }
    }

    let monthActiveIndex = 0
    let bodyScrollTop = 0

    function onScroll(e: Event) {
      bodyScrollTop = getScrollTop(e.currentTarget as HTMLDivElement)

      updateBodyFixed(bodyScrollTop)
    }

    function updateBodyTitle(t: string, tY: number | null) {
      if (!bodyTitleEl.value) {
        return
      }

      bodyTitleEl.value.textContent = t
      bodyTitleEl.value.style.cssText = CSSProperties2CssText(
        getViewBodyTitleStyles(tY)
      )
    }

    function updateBodyFixed(scrollTop: number) {
      const h = 28
      const $items: HTMLDivElement[] = bodyEl.value
        ? [].slice.call(
            bodyEl.value.querySelectorAll('.fx-virtual-list_item'),
            0
          )
        : []

      function getItemName(vIndex: number) {
        const realIndex = $items[vIndex]
          ? parseInt($items[vIndex].dataset.index as string)
          : -1

        return realIndex === -1 ? '' : months[realIndex].caption
      }

      if ($items.length === 0) {
        updateBodyTitle('', null)
        return
      }

      const _index = monthActiveIndex
      const nextIndex = _index + 1
      const offsetTops = $items.map($el => {
        const matches = $el.style.cssText.match(
          /translate3d\(\w+,\s*(\d+)px,\s*\w+\)/
        )

        return matches && matches[1] ? parseFloat(matches[1]) : -1
      })

      const current = offsetTops[_index]
      const next =
        offsetTops[nextIndex] != null ? offsetTops[nextIndex] : Infinity
      const first = offsetTops[0]

      if (scrollTop < first) {
        updateBodyTitle('', null)
      } else if (scrollTop >= current) {
        if (scrollTop >= next) {
          monthActiveIndex = nextIndex
          updateBodyTitle(getItemName(nextIndex), 0)

          if (
            offsetTops[nextIndex + 1] &&
            scrollTop >= offsetTops[nextIndex + 1]
          ) {
            // 超过了
            updateBodyFixed(scrollTop)
          }
        } else if (next - scrollTop < h) {
          updateBodyTitle(getItemName(_index), next - scrollTop - h)
        } else {
          updateBodyTitle(getItemName(_index), 0)
        }
      } else {
        if (current - scrollTop < h) {
          updateBodyTitle(getItemName(_index - 1), current - scrollTop - h)
        } else {
          monthActiveIndex = _index - 1
          updateBodyTitle(getItemName(_index - 1), 0)

          if (offsetTops[_index - 1] && offsetTops[_index - 1] > scrollTop) {
            updateBodyFixed(scrollTop)
          }
        }
      }
    }

    function getItemSize(index: number) {
      return Math.ceil((months[index]?.days.length ?? 0) / 7) * 64 + 28
    }

    watch(
      [() => props.minDate, () => props.maxDate, locale],
      () => {
        updateOptions()
        updateOriginalValue([start.timestamp, end.timestamp])
      },
      {
        deep: true,
        immediate: true
      }
    )

    watch(
      () => props.modelValue,
      val => updateValue(val),
      {
        deep: true,
        immediate: true
      }
    )

    onMounted(() => {
      updateBodyFixed(bodyScrollTop)
    })

    return {
      mode,
      weekDays,
      months,
      onDaysClick,
      getDetail,
      locale,
      getItemSize,
      bodyEl,
      onScroll,
      bodyTitleEl
    }
  }
})
</script>
