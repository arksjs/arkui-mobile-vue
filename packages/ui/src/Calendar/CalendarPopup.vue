<template>
  <Drawer
    class="ta-calendar-popup"
    placement="bottom"
    :visible="visible"
    :title="title"
    :showClose="showClose"
    @visibleStateChange="onVisibleStateChange"
    @confirm="onConfirm"
    @cancel="onCancel"
    @update:visible="onUpdateVisible"
    ref="popupRef"
  >
    <CalendarView
      :modelValue="modelValue"
      :minDate="minDate"
      :maxDate="maxDate"
      :initialMode="initialMode"
      :allowSameDay="allowSameDay"
      :maxRange="maxRange"
      :dayHandler="dayHandler"
      :firstDayOfWeek="firstDayOfWeek"
      :formatter="formatter"
      :parser="parser"
      ref="viewRef"
      @select="onViewSelect"
    />
    <div class="ta-calendar-popup_confirm" v-if="showConfirm">
      <TaButton type="primary" size="large" @click="onConfirmClick" :disabled="valueSize == 0"
        >{{ locale.calendarConfirmText }}
      </TaButton>
    </div>
  </Drawer>
</template>

<script lang="ts">
import { defineComponent, nextTick, onMounted, ref, shallowRef, watch } from 'vue'
import CalendarView from './CalendarView.vue'
import { Drawer } from '../Drawer'
import { Button as TaButton } from '../Button'
import { commonProps, calendarDetailValidator } from './props'
import { usePopupExtend } from '../popup/use-popup'
import { popupExtendProps } from '../popup/props'
import { pickerPopupEmits } from '../Picker/props'
import { cloneDetail, isSameDetail, isValidValue } from '../Picker/util'
import type {
  CalendarSelectorDetail,
  CalendarPopupEmits,
  CalendarViewRef,
  CalendarDetail
} from './types'
import { useLocale } from '../ConfigProvider/context'
import { useHandlers } from './use-calendar'
import { cloneData, type PropsToEmits } from '../helpers'
import { getSourceDetail } from './util'

export default defineComponent({
  name: 'ta-calendar-popup',
  components: { CalendarView, Drawer, TaButton },
  props: {
    ...commonProps,
    ...popupExtendProps,
    title: {
      type: String
    },
    showConfirm: {
      type: Boolean,
      default: false
    },
    showClose: {
      type: Boolean,
      default: false
    }
  },
  emits: {
    ...pickerPopupEmits,
    confirm: calendarDetailValidator
  } as PropsToEmits<CalendarPopupEmits>,
  setup(props, ctx) {
    const { emit } = ctx
    const { locale } = useLocale()
    const viewRef = shallowRef<CalendarViewRef | null>(null)
    const valueSize = ref(0)

    const { getDefaultDetail } = useHandlers(props)

    let detail: CalendarSelectorDetail = getDefaultDetail()

    const popup = usePopupExtend<CalendarDetail>(ctx)

    function onViewSelect() {
      const newDetail = getViewDetail()

      valueSize.value = newDetail.valueArray.length

      if (!props.showConfirm) {
        confirm()
      } else {
        // 判断下是否可以confirm
        valueSize.value = newDetail.valueArray.length
      }
    }

    function onConfirmClick() {
      confirm()
    }

    function confirm() {
      const newDetail = getViewDetail()

      if (!isSameDetail(newDetail, detail)) {
        updateDetail(newDetail)

        emit('update:modelValue', getDetail().value)
        const { value, label } = getDetail()
        emit('change', value, label)
      } else {
        updateDetail(newDetail)
      }

      popup.customConfirm(getSourceDetail(getDetail()))
    }

    function getDetail() {
      return cloneDetail(detail)
    }

    function updateDetail(newDetail: CalendarSelectorDetail) {
      detail = newDetail
      valueSize.value = newDetail.valueArray.length
    }

    function getViewDetail() {
      return viewRef.value?.getDetail() || getDefaultDetail()
    }

    if (isValidValue(props.modelValue)) {
      // 如果默认设置了modelValue，先缓存一下
      detail.value = cloneData(props.modelValue)
    }

    watch(
      () => props.modelValue,
      () => {
        nextTick(() => {
          detail = getViewDetail()
        })
      },
      {
        deep: true
      }
    )

    onMounted(() => {
      detail = getViewDetail()
    })

    return {
      ...popup,
      valueSize,
      viewRef,
      onViewSelect,
      onConfirmClick,
      getDetail,
      locale
    }
  }
})
</script>
