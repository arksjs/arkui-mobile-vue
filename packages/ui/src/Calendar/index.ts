import Calendar from './Calendar.vue'
import CalendarPopup from './CalendarPopup.vue'
import CalendarView from './CalendarView.vue'
import { createConfirmHook, createShowPopup } from '../popup/api'
import type { PopupSuccessConfirmArgs } from '../popup/types'
import type { ShowCalendarOptions, CalendarDetail } from './types'

const showCalendar = createShowPopup<ShowCalendarOptions, PopupSuccessConfirmArgs<CalendarDetail>>({
  apiName: 'showCalendar',
  component: CalendarPopup,
  createHook: createConfirmHook
})

export { Calendar, CalendarPopup, CalendarView, showCalendar }
export default Calendar
