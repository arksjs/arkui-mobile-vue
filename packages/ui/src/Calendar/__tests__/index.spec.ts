import { describe, test, expect } from 'vitest'
import { shallowMount } from '@vue/test-utils'
import { Calendar, CalendarPopup, CalendarView } from '@/Calendar'

describe('Calendar', () => {
  test.concurrent('snapshot', () => {
    const wrapper = shallowMount(Calendar, {
      props: {
        options: []
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('CalendarPopup', () => {
  test.concurrent('snapshot', () => {
    const wrapper = shallowMount(CalendarPopup, {
      props: {
        options: []
      },
      global: {
        stubs: { teleport: true }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('CalendarView', () => {
  test.concurrent('snapshot', () => {
    const wrapper = shallowMount(CalendarView, {
      props: {
        options: []
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
