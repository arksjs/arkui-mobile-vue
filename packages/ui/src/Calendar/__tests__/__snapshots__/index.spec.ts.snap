// Vitest Snapshot v1

exports[`Calendar > snapshot 1`] = `
"<div class=\\"ta-calendar\\">
  <selector-field-stub label=\\"\\" name=\\"\\" placeholder=\\"\\" disabled=\\"false\\" value=\\"\\"></selector-field-stub>
  <!--v-if-->
</div>"
`;

exports[`CalendarPopup > snapshot 1`] = `"<drawer-stub visible=\\"false\\" maskclosable=\\"true\\" placement=\\"bottom\\" showclose=\\"false\\" enablesafeareainsets=\\"true\\" showmask=\\"true\\" class=\\"ta-calendar-popup\\"></drawer-stub>"`;

exports[`CalendarView > snapshot 1`] = `
"<div class=\\"ta-calendar-view\\">
  <div class=\\"ta-calendar-view_header\\">
    <div class=\\"ta-calendar-view_weekdays\\"><span class=\\"ta-calendar-view_weekday highlight\\">日</span><span class=\\"ta-calendar-view_weekday\\">一</span><span class=\\"ta-calendar-view_weekday\\">二</span><span class=\\"ta-calendar-view_weekday\\">三</span><span class=\\"ta-calendar-view_weekday\\">四</span><span class=\\"ta-calendar-view_weekday\\">五</span><span class=\\"ta-calendar-view_weekday highlight\\">六</span></div>
  </div>
  <div class=\\"ta-calendar-view_body\\">
    <virtual-list-stub itemsize=\\"function getItemSize(index) {
      return Math.ceil((months[index]?.days.length ?? 0) / 7) * 64 + 28;
    }\\" ids=\\"2022年01月,2022年02月,2022年03月,2022年04月,2022年05月,2022年06月,2022年07月,2022年08月,2022年09月,2022年10月,2022年11月,2022年12月\\" initialhorizontal=\\"false\\" preload=\\"1.5\\" initialwaterfallcount=\\"1\\" approveditemvisiblescale=\\"0.5\\"></virtual-list-stub>
    <div class=\\"ta-calendar-view_month-caption fixed\\" style=\\"transform: translate3d(0px, -100%, 0px);\\"></div>
  </div>
</div>"
`;
